app/gather/api/utils.py

```diff
@@ -0,0 +1,252 @@
+        url=url,
+        json=data or {},
+        headers=headers,
+    )
+    try:
+        res.raise_for_status()
+    except HTTPError as e:
+        LOG.debug(str(e))
+        raise e
+    try:
+        return res.json()
+    except Exception:
+        return res.content
+
+
+def upsert_resource(base_url, resource_name, resource, headers={}, skip_subs=False):
+    '''
+    Upsert consumer resources
+    '''
+    _resource_id = resource.get('id')
+    if _resource_id:
+        try:
+            _existing_resource = consumer_request(
+                url=f'{base_url}{resource_name}/list',
+                headers=headers
+            )
+            _method = f'update?id={_resource_id}' if _resource_id in _existing_resource else 'add'
+            url = f'{base_url}{resource_name}/{_method}'
+
+            if resource_name == 'job' and _method != 'add' and not skip_subs:
+                _existing_job = consumer_request(
+                    url=f'{base_url}job/get?id={_resource_id}',
+                    headers=headers
+                )
+                old_subs = _existing_job.get('subscription', [])
+                new_subs = list(set(old_subs + resource['subscription']))
+                resource['subscription'] = new_subs
+
+            return consumer_request(
+                url=url,
+                method='post',
+                headers=headers,
+                data=resource
+            )
+        except Exception as e:
+            LOG.debug(str(e))
+            raise e
+    else:
+        err = f'{resource_name}: No id provided'
+        LOG.debug(err)
+        raise Exception(err)
+
+
+def __clean_name(value):
+    '''
+    Replaces any non alphanumeric character with spaces
+    Converts to title case
+    Removes spaces
+    '''
+    return ''.join([c if c.isalnum() else ' ' for c in value]).title().replace(' ', '')
+
+
+def configure_consumers(consumer_settings, survey_name, headers={}):
+    '''
+    Creates or updates consumer resources
+    Add a subscription to a topic as the survey name
+
+    Returns list of consumers configured and list of errors if any
+    '''
+
+    cons_errors = []
+    cons = []
+    consumer_env_variables = {}
+
+    for consumer in consumer_settings:
+        if not consumer.keys() >= {'resources', 'subscription', 'job', 'url', 'name'}:
+            cons_errors += [
+                {consumer.get('name', 'Unknown'): 'Invalid config settings'}
+            ]
+            continue
+
+        consumer_name = consumer.pop('name')
+        consumer_url = consumer.pop('url')
+        consumer_resources = consumer.pop('resources')
+        job_data = {}
+        consumer_env_variables[consumer_name] = get_env_variables(consumer_name)
+
+        # check if consumer is online
+        try:
+            consumer_request(url=f'{consumer_url}health', headers=headers)
+        except Exception as e:
+            cons_errors += [
+                {consumer_name: 'Consumer is offline'}
```

Comment:
to translate or not to translate, that's the question

Reply:
to translate